File: host/meshctl/compat.py
"""INSECURE-only receive parser for Meshtastic over-the-air frames.

INSECURE: Meshtastic compat traffic uses a shared channel PSK that
decrypts all traffic, has no forward secrecy, floods up to 7 hops, and
may leak via MQTT bridging. Anything decoded by this module MUST be
treated as public, unauthenticated, untrusted display text only.

Scope (receive-parse ONLY):

- Decode a raw LoRa payload captured by a SECOND radio tuned to
  Meshtastic parameters into sender/destination/portnum/text for
  display.
- NEVER encrypt, encode, or send as Meshtastic: this module has no
  ``encode``/``send``/``encrypt`` path by design.
- NEVER handle channel PSKs and NEVER import keys into the secure
  mesh path. Packets using channel encryption (``encrypted`` field) are
  reported as undecodable-by-design; no key API exists here.
- NEVER touch the private secure mesh protocol (LMESH framing, X25519
  pairing, ChaCha20-Poly1305 traffic keys). No import from any secure
  module; stdlib only.

Caller contract: pass the exact LoRa payload bytes delivered by the
second (Meshtastic-tuned) radio. RF capture, serial transport, and any
forwarding policy live outside this module.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field

__all__ = [
    "BROADCAST_ADDR",
    "MAX_FRAME_LEN",
    "MeshtasticParseError",
    "MeshtasticData",
    "MeshtasticPacket",
    "portnum_name",
    "parse_packet",
    "try_parse_packet",
    "format_packet",
]

#: INSECURE: Meshtastic broadcast destination (``^all`` / ``0xFFFFFFFF``).
BROADCAST_ADDR = 0xFFFFFFFF

#: INSECURE: largest LoRa payload this parser accepts. Oversize input is
#: rejected, never truncated-and-decoded.
MAX_FRAME_LEN = 512

#: INSECURE: known Meshtastic ``portnum`` values (subset used for display
#: labels; unknown values are reported as ``UNKNOWN(<n>)`` and dropped
#: to a length summary, never executed/dispatched).
_PORTNUM_NAMES = {
    0: "UNKNOWN_APP",
    1: "TEXT_MESSAGE_APP",
    2: "REMOTE_HARDWARE_APP",
    3: "POSITION_APP",
    4: "NODEINFO_APP",
    5: "ROUTING_APP",
    6: "ADMIN_APP",
    7: "TEXT_MESSAGE_COMPRESSED_APP",
    8: "WAYPOINT_APP",
    9: "AUDIO_APP",
    10: "DETECTION_SENSOR_APP",
    11: "ALERT_APP",
    12: "KEY_VERIFICATION_APP",
    32: "REPLY_APP",
    33: "IP_TUNNEL_APP",
    64: "SERIAL_APP",
    65: "STORE_FORWARD_APP",
    66: "RANGE_TEST_APP",
    67: "TELEMETRY_APP",
    68: "ZPS_APP",
    69: "SIMULATOR_APP",
    70: "TRACEROUTE_APP",
    71: "NEIGHBORINFO_APP",
    72: "ATAK_PLUGIN",
    73: "MAP_REPORT_APP",
    74: "POWERSTRESS_APP",
    256: "PRIVATE_APP",
    257: "ATAK_FORWARDER",
}

_TEXT_PORT = 1
_POSITION_PORT = 3


class MeshtasticParseError(ValueError):
    """INSECURE: raised when raw bytes are not a decodable public packet."""


@dataclass(frozen=True)
class MeshtasticData:
    """INSECURE: decoded ``Data`` envelope of one public packet."""

    portnum: int
    portname: str
    payload: bytes
    text: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    strings: tuple[str, ...] = ()


@dataclass(frozen=True)
class MeshtasticPacket:
    """INSECURE: one decoded public Meshtastic ``MeshPacket``."""

    from_id: int = 0
    to_id: int = 0
    channel: int = 0
    packet_id: int = 0
    hop_limit: int = 0
    hop_start: int = 0
    want_ack: bool = False
    via_mqtt: bool = False
    rx_time: int = 0
    rx_snr: float | None = None
    rx_rssi: int | None = None
    data: MeshtasticData | None = None
    encrypted: bool = False
    encrypted_len: int = 0
    unknown_fields: tuple[int, ...] = field(default=())


def portnum_name(portnum: int) -> str:
    """INSECURE: display label for a Meshtastic ``portnum`` value."""
    return _PORTNUM_NAMES.get(portnum, f"UNKNOWN({portnum})")


def parse_packet(raw: bytes | bytearray) -> MeshtasticPacket:
    """INSECURE: parse one raw LoRa payload into a public packet.

    Only unencrypted (``decoded``) packets yield ``data``. Packets
    carrying the ``encrypted`` field yield ``encrypted=True`` with no
    payload access — there is deliberately no key/decrypt API in this
    module. Raises :class:`MeshtasticParseError` on empty, oversize, or
    malformed input.
    """
    buf = bytes(raw)
    if not buf:
        raise MeshtasticParseError("empty frame")
    if len(buf) > MAX_FRAME_LEN:
        raise MeshtasticParseError(f"frame too long: {len(buf)} bytes")

    from_id = 0
    to_id = 0
    channel = 0
    packet_id = 0
    hop_limit = 0
    hop_start = 0
    want_ack = False
    via_mqtt = False
    rx_time = 0
    rx_snr: float | None = None
    rx_rssi: int | None = None
    data: MeshtasticData | None = None
    encrypted = False
    encrypted_len = 0
    unknown: list[int] = []
    seen_any = False

    for field_no, wire, value in _iter_fields(buf):
        seen_any = True
        if field_no == 1 and wire == 5:
            from_id = value
        elif field_no == 2 and wire == 5:
            to_id = value
        elif field_no == 3 and wire == 0:
            channel = value
        elif field_no == 4 and wire == 2:
            if data is not None:
                raise MeshtasticParseError("duplicate decoded field")
            data = _parse_data(value)
        elif field_no == 5 and wire == 2:
            encrypted = True
            encrypted_len = len(value)
        elif field_no == 6 and wire == 5:
            packet_id = value
        elif field_no == 7 and wire == 5:
            rx_time = value
        elif field_no == 8 and wire == 5:
            (rx_snr,) = struct.unpack("<f", value.to_bytes(4, "little"))
        elif field_no == 9 and wire == 0:
            hop_limit = value
        elif field_no == 10 and wire == 0:
            want_ack = bool(value)
        elif field_no == 12 and wire == 0:
            rx_rssi = _to_int32(value)
        elif field_no == 14 and wire == 0:
            via_mqtt = bool(value)
        elif field_no == 15 and wire == 0:
            hop_start = value
        else:
            unknown.append(field_no)

    if not seen_any:
        raise MeshtasticParseError("no protobuf fields found")
    if data is not None and encrypted:
        raise MeshtasticParseError("packet has both decoded and encrypted")
    if data is None and not encrypted:
        raise MeshtasticParseError("packet has neither decoded nor encrypted")

    return MeshtasticPacket(
        from_id=from_id,
        to_id=to_id,
        channel=channel,
        packet_id=packet_id,
        hop_limit=hop_limit,
        hop_start=hop_start,
        want_ack=want_ack,
        via_mqtt=via_mqtt,
        rx_time=rx_time,
        rx_snr=rx_snr,
        rx_rssi=rx_rssi,
        data=data,
        encrypted=encrypted,
        encrypted_len=encrypted_len,
        unknown_fields=tuple(unknown),
    )


def try_parse_packet(raw: bytes | bytearray) -> MeshtasticPacket | None:
    """INSECURE: like :func:`parse_packet` but returns ``None`` instead.

    Returns ``None`` (never raises) for empty/oversize/malformed input
    so a TUI receive loop can count-and-drop undecodable RF noise.
    """
    try:
        return parse_packet(raw)
    except (MeshtasticParseError, ValueError, struct.error, IndexError):
        return None


def format_packet(pkt: MeshtasticPacket) -> str:
    """INSECURE: render a parsed packet as one display line.

    Output is untrusted display text: the ``INSECURE`` prefix and the
    ``(shared channel; unauthenticated)`` marker MUST be preserved by
    any UI that shows this string.
    """
    head = (
        f"INSECURE meshtastic from={pkt.from_id:#010x} "
        f"to={pkt.to_id:#010x} ch={pkt.channel} "
        f"id={pkt.packet_id:#010x} "
        f"hop={pkt.hop_start}/{pkt.hop_limit} "
        f"(shared channel; unauthenticated)"
    )
    if pkt.via_mqtt:
        head += " [via-mqtt]"
    if pkt.encrypted or pkt.data is None:
        return f"{head} encrypted packet ({pkt.encrypted_len} B; not decrypted)"
    data = pkt.data
    if data.portnum == _TEXT_PORT and data.text is not None:
        return f"{head} {data.portname}: {data.text}"
    if data.portnum == _POSITION_PORT and (
        data.latitude is not None or data.longitude is not None
    ):
        return f"{head} {data.portname}: lat={data.latitude} lon={data.longitude}"
    if data.strings:
        preview = " | ".join(data.strings[:4])
        return f"{head} {data.portname} ({len(data.payload)} B): {preview}"
    return f"{head} {data.portname} ({len(data.payload)} B)"


def _parse_data(buf: bytes) -> MeshtasticData:
    """Decode a ``Data`` protobuf envelope (internal helper)."""
    portnum = 0
    payload = b""
    for field_no, wire, value in _iter_fields(buf):
        if field_no == 1 and wire == 0:
            portnum = value
        elif field_no == 2 and wire == 2:
            if payload:
                raise MeshtasticParseError("duplicate Data payload")
            payload = value
        # All other Data fields (want_response, dest, source,
        # request/reply ids, emoji, bitfield) are skipped: display only.

    name = portnum_name(portnum)
    text: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    strings: tuple[str, ...] = ()

    if portnum == _TEXT_PORT:
        text = _decode_text(payload)
    elif portnum == _POSITION_PORT:
        latitude, longitude = _decode_position(payload)
    elif payload:
        strings = _harvest_strings(payload)

    return MeshtasticData(
        portnum=portnum,
        portname=name,
        payload=payload,
        text=text,
        latitude=latitude,
        longitude=longitude,
        strings=strings,
    )


def _decode_text(payload: bytes) -> str:
    """Decode a TEXT_MESSAGE_APP payload as UTF-8 (internal helper)."""
    if not payload:
        raise MeshtasticParseError("empty text payload")
    if len(payload) > MAX_FRAME_LEN:
        raise MeshtasticParseError("text payload too long")
    return payload.decode("utf-8", errors="replace")


def _decode_position(payload: bytes) -> tuple[float | None, float | None]:
    """Best-effort lat/lon from a POSITION_APP payload (internal helper).

    Only ``latitude_i`` (field 1, sfixed32) and ``longitude_i``
    (field 2, sfixed32), scaled by 1e7, are extracted. ``(None, None)``
    is returned when neither field is present; other position fields
    are ignored.
    """
    latitude: float | None = None
    longitude: float | None = None
    for field_no, wire, value in _iter_fields(payload):
        if wire != 5:
            continue
        raw = struct.unpack("<i", value.to_bytes(4, "little"))[0]
        if field_no == 1 and latitude is None:
            latitude = raw / 1e7
        elif field_no == 2 and longitude is None:
            longitude = raw / 1e7
    return latitude, longitude


def _harvest_strings(payload: bytes, limit: int = 64) -> tuple[str, ...]:
    """Collect printable ASCII runs for display of opaque payloads.

    Heuristic only (internal helper): used to preview NODEINFO-style
    payloads without implementing their full protobuf schema. Runs
    shorter than 3 characters are ignored; at most ``limit`` total
    characters are returned across all runs.
    """
    runs: list[str] = []
    current: list[str] = []
    budget = limit

    def flush() -> None:
        nonlocal budget
        if len(current) >= 3 and budget > 0:
            run = "".join(current)[:budget]
            runs.append(run)
            budget -= len(run)
        current.clear()

    for byte in payload:
        if 32 <= byte < 127:
            current.append(chr(byte))
        else:
            flush()
    flush()
    return tuple(runs)


def _iter_fields(buf: bytes) -> object:
    """Yield ``(field_number, wire_type, value)`` protobuf fields.

    Internal generator. ``value`` is an ``int`` for varint (wire 0) and
    fixed32 (wire 5, little-endian unsigned), ``bytes`` for
    length-delimited (wire 2); 64-bit (wire 1) values are skipped. Any
    truncation or illegal tag raises :class:`MeshtasticParseError`.
    """
    pos = 0
    end = len(buf)
    while pos < end:
        tag, pos = _read_varint(buf, pos)
        field_no = tag >> 3
        wire = tag & 0x07
        if field_no == 0:
            raise MeshtasticParseError("bad protobuf tag")
        if wire == 0:
            value, pos = _read_varint(buf, pos)
            yield field_no, wire, value
        elif wire == 1:
            if pos + 8 > end:
                raise MeshtasticParseError("truncated 64-bit field")
            pos += 8  # RX metadata; value unused for display.
        elif wire == 2:
            length, pos = _read_varint(buf, pos)
            if pos + length > end:
                raise MeshtasticParseError("truncated length-delimited field")
            value = buf[pos : pos + length]
            pos += length
            yield field_no, wire, value
        elif wire == 5:
            if pos + 4 > end:
                raise MeshtasticParseError("truncated 32-bit field")
            value = int.from_bytes(buf[pos : pos + 4], "little")
            pos += 4
            yield field_no, wire, value
        else:
            raise MeshtasticParseError(f"unsupported wire type {wire}")


def _read_varint(buf: bytes, pos: int) -> tuple[int, int]:
    """Read one protobuf varint at ``pos`` (internal helper)."""
    result = 0
    shift = 0
    while True:
        if pos >= len(buf):
            raise MeshtasticParseError("truncated varint")
        byte = buf[pos]
        pos += 1
        result |= (byte & 0x7F) << shift
        if not byte & 0x80:
            return result, pos
        shift += 7
        if shift >= 70:
            raise MeshtasticParseError("varint overflow")


def _to_int32(value: int) -> int:
    """Interpret a protobuf int32 varint as signed (internal helper)."""
    value &= (1 << 64) - 1
    if value >= 1 << 63:
        value -= 1 << 64
    if value < -(1 << 31) or value >= 1 << 31:
        raise MeshtasticParseError("int32 out of range")
    return value

File: crates/mesh-core/src/compat.rs
//! INSECURE-only receive parser for Meshtastic over-the-air frames.
//!
//! INSECURE: Meshtastic compat traffic uses a shared channel PSK that
//! decrypts all traffic, has no forward secrecy, floods up to 7 hops, and
//! may leak via MQTT bridging. Anything decoded by this module MUST be
//! treated as public, unauthenticated, untrusted display bytes only.
//!
//! Scope (decode ONLY, behind the non-default `meshtastic` cargo feature):
//!
//! - Decode a raw Meshtastic `MeshPacket` payload (protobuf, `decoded`
//!   form only) into sender/destination/portnum bytes for display.
//! - NEVER encrypt, encode, or send as Meshtastic: this module has no
//!   `encode`/`send`/`encrypt` path by design.
//! - NEVER handle channel PSKs. Packets carrying the `encrypted` field
//!   yield `encrypted: true` with no payload access — there is
//!   deliberately no key/decrypt API here.
//! - NEVER touch the private secure mesh protocol (LMESH framing, X25519
//!   pairing, ChaCha20-Poly1305 traffic keys). No import from any secure
//!   module; `core` only.
//!
//! Caller contract (encrypted-path-only): compat bytes ride INSIDE the
//! existing encrypted LMESH DATA body and are handed to
//! [`parse_packet`] only after endpoint decryption. This module never
//! sees raw RF, never retunes the modem, and never bypasses the secure
//! frame path. RF capture (second, Meshtastic-tuned radio) and any
//! forwarding policy live outside this module.

use core::str;

/// INSECURE: Meshtastic broadcast destination (`^all` / `0xFFFFFFFF`).
pub const BROADCAST_ADDR: u32 = 0xFFFF_FFFF;

/// INSECURE: largest LoRa payload this parser accepts. Oversize input is
/// rejected, never truncated-and-decoded.
pub const MAX_COMPAT_FRAME_LEN: usize = 512;

/// INSECURE: text-message portnum (`TEXT_MESSAGE_APP`).
pub const PORT_TEXT: u32 = 1;
/// INSECURE: position portnum (`POSITION_APP`).
pub const PORT_POSITION: u32 = 3;

/// Decode failure modes. No payload data is exposed on error.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CompatError {
    /// Input was empty.
    Empty,
    /// Input exceeded [`MAX_COMPAT_FRAME_LEN`].
    TooLong,
    /// No protobuf fields found.
    NoFields,
    /// Buffer ended mid-field.
    Truncated,
    /// Illegal protobuf tag (field number 0).
    BadTag,
    /// Varint exceeded 10 bytes / overflowed u64.
    VarintOverflow,
    /// Unsupported protobuf wire type (3, 4, 6, 7).
    UnsupportedWire,
    /// Duplicate `decoded` field or duplicate `Data` payload.
    Duplicate,
    /// Packet carried both `decoded` and `encrypted`.
    BothDecodedAndEncrypted,
    /// Packet carried neither `decoded` nor `encrypted`.
    NeitherDecodedNorEncrypted,
    /// Text payload is not valid UTF-8.
    BadUtf8,
    /// Text payload was empty.
    TextEmpty,
    /// Signed int32 varint out of range.
    Int32OutOfRange,
}

/// INSECURE: decoded `Data` envelope of one public packet. `payload` borrows
/// the input buffer; no allocation, no copy.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CompatData<'a> {
    /// Meshtastic `portnum` value.
    pub portnum: u32,
    /// Raw `payload` bytes (borrowed).
    pub payload: &'a [u8],
}

impl<'a> CompatData<'a> {
    /// INSECURE: interpret the payload as a text message. Validates UTF-8
    /// strictly (no lossy replacement); rejects empty payloads.
    pub fn text(&self) -> Result<&'a str, CompatError> {
        if self.payload.is_empty() {
            return Err(CompatError::TextEmpty);
        }
        str::from_utf8(self.payload).map_err(|_| CompatError::BadUtf8)
    }

    /// Best-effort `(latitude, longitude)` in degrees from a `POSITION_APP`
    /// payload. Only `latitude_i` (field 1, sfixed32) and `longitude_i`
    /// (field 2, sfixed32), scaled by 1e7, are extracted; `(None, None)`
    /// when neither is present. Other position fields are ignored.
    pub fn position(&self) -> (Option<f32>, Option<f32>) {
        let mut latitude: Option<f32> = None;
        let mut longitude: Option<f32> = None;
        let mut fields = ProtoFields::new(self.payload);
        loop {
            match fields.next() {
                None => break,
                Some(Err(_)) => break,
                Some(Ok((field_no, wire, value))) => {
                    if wire != Wire::Fixed32 {
                        continue;
                    }
                    let bits = match value {
                        FieldVal::Fixed32(b) => b,
                        _ => continue,
                    };
                    #[allow(clippy::cast_possible_wrap)]
                    let raw = bits as i32;
                    // f32 has 24-bit mantissa; 1e7-scaled i32 degrees lose
                    // sub-meter precision here — display-only, acceptable.
                    #[allow(clippy::cast_precision_loss)]
                    let deg = raw as f32 / 10_000_000.0;
                    if field_no == 1 && latitude.is_none() {
                        latitude = Some(deg);
                    } else if field_no == 2 && longitude.is_none() {
                        longitude = Some(deg);
                    }
                }
            }
        }
        (latitude, longitude)
    }
}

/// INSECURE: one decoded public Meshtastic `MeshPacket`. All slices borrow
/// the input buffer; `unknown_count` saturates at 255.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CompatPacket<'a> {
    /// `from` node id (fixed32 field 1).
    pub from_id: u32,
    /// `to` node id (fixed32 field 2).
    pub to_id: u32,
    /// Channel index (varint field 3).
    pub channel: u64,
    /// Packet id (fixed32 field 6).
    pub packet_id: u32,
    /// Remaining hop limit (varint field 9).
    pub hop_limit: u64,
    /// Original hop count (varint field 15).
    pub hop_start: u64,
    /// `want_ack` flag (varint field 10).
    pub want_ack: bool,
    /// `via_mqtt` flag (varint field 14).
    pub via_mqtt: bool,
    /// RX timestamp (fixed32 field 7; sender-local, unauthenticated).
    pub rx_time: u32,
    /// Raw RX SNR float bits (fixed32 field 8), if present.
    pub rx_snr_bits: Option<u32>,
    /// RX RSSI in dBm (varint field 12, int32), if present.
    pub rx_rssi: Option<i32>,
    /// Decoded `Data` envelope; `None` for channel-encrypted packets.
    pub data: Option<CompatData<'a>>,
    /// True when the packet carries the `encrypted` field (opaque by design).
    pub encrypted: bool,
    /// Length of the opaque `encrypted` bytes (0 for `decoded` packets).
    pub encrypted_len: usize,
    /// Count of skipped unknown field numbers (saturates at 255).
    pub unknown_count: u8,
}

/// INSECURE: display label for a Meshtastic `portnum` value. Unknown values
/// are display-only names, never dispatched.
#[must_use]
pub fn portnum_name(portnum: u32) -> &'static str {
    match portnum {
        0 => "UNKNOWN",
        1 => "TEXT_MESSAGE_APP",
        2 => "REMOTE_HARDWARE_APP",
        3 => "POSITION_APP",
        4 => "NODEINFO_APP",
        5 => "ROUTING_APP",
        6 => "ADMIN_APP",
        7 => "TEXT_MESSAGE_COMPRESSED_APP",
        8 => "WAYPOINT_APP",
        9 => "AUDIO_APP",
        10 => "DETECTION_SENSOR_APP",
        11 => "ALERT_APP",
        12 => "KEY_VERIFICATION_APP",
        13 => "REPLY_APP",
        14 => "IP_TUNNEL_APP",
        15 => "SERIAL_APP",
        16 => "STORE_FORWARD_APP",
        17 => "RANGE_TEST_APP",
        18 => "TELEMETRY_APP",
        19 => "ZPS_APP",
        20 => "ATAK_PLUGIN",
        21 => "MAP_REPORT_APP",
        22 => "POWERSTRESS_APP",
        32 => "PRIVATE_APP",
        33 => "ATAK_FORWARDER",
        64 => "SIMULATOR_APP",
        256 => "TRACEROUTE_APP",
        257 => "NEIGHBORINFO_APP",
        _ => "UNKNOWN",
    }
}

/// INSECURE: parse one raw Meshtastic payload into a public packet.
///
/// Only unencrypted (`decoded`) packets yield `data`. Packets carrying the
/// `encrypted` field yield `encrypted: true` with no payload access.
/// Returns [`CompatError`] on empty, oversize, or malformed input.
pub fn parse_packet(raw: &[u8]) -> Result<CompatPacket<'_>, CompatError> {
    if raw.is_empty() {
        return Err(CompatError::Empty);
    }
    if raw.len() > MAX_COMPAT_FRAME_LEN {
        return Err(CompatError::TooLong);
    }

    let mut pkt = CompatPacket {
        from_id: 0,
        to_id: 0,
        channel: 0,
        packet_id: 0,
        hop_limit: 0,
        hop_start: 0,
        want_ack: false,
        via_mqtt: false,
        rx_time: 0,
        rx_snr_bits: None,
        rx_rssi: None,
        data: None,
        encrypted: false,
        encrypted_len: 0,
        unknown_count: 0,
    };
    let mut fields = ProtoFields::new(raw);
    while let Some(item) = fields.next() {
        let (field_no, wire, value) = item?;
        match (field_no, wire) {
            (1, Wire::Fixed32) => pkt.from_id = fixed32(value),
            (2, Wire::Fixed32) => pkt.to_id = fixed32(value),
            (3, Wire::Varint) => pkt.channel = varint(value),
            (4, Wire::Len) => {
                if pkt.data.is_some() {
                    return Err(CompatError::Duplicate);
                }
                pkt.data = Some(parse_data(len_bytes(value))?);
            }
            (5, Wire::Len) => {
                pkt.encrypted = true;
                pkt.encrypted_len = len_bytes(value).len();
            }
            (6, Wire::Fixed32) => pkt.packet_id = fixed32(value),
            (7, Wire::Fixed32) => pkt.rx_time = fixed32(value),
            (8, Wire::Fixed32) => pkt.rx_snr_bits = Some(fixed32(value)),
            (9, Wire::Varint) => pkt.hop_limit = varint(value),
            (10, Wire::Varint) => pkt.want_ack = varint(value) != 0,
            (12, Wire::Varint) => pkt.rx_rssi = Some(varint_to_int32(varint(value))?),
            (14, Wire::Varint) => pkt.via_mqtt = varint(value) != 0,
            (15, Wire::Varint) => pkt.hop_start = varint(value),
            // 64-bit values are RX metadata, skipped silently (never stored).
            (_, Wire::Fixed64) => {}
            _ => {
                pkt.unknown_count = pkt.unknown_count.saturating_add(1);
            }
        }
    }
    if pkt.data.is_some() && pkt.encrypted {
        return Err(CompatError::BothDecodedAndEncrypted);
    }
    if pkt.data.is_none() && !pkt.encrypted {
        return Err(CompatError::NeitherDecodedNorEncrypted);
    }
    Ok(pkt)
}

/// INSECURE: like [`parse_packet`] but returns `None` instead of an error,
/// so a receive loop can count-and-drop undecodable RF noise.
#[must_use]
pub fn try_parse_packet(raw: &[u8]) -> Option<CompatPacket<'_>> {
    parse_packet(raw).ok()
}

/// Decode a `Data` protobuf envelope (internal helper). All non-payload
/// `Data` fields (want_response, dest, source, request/reply ids, emoji,
/// bitfield) are skipped: display only.
fn parse_data(buf: &[u8]) -> Result<CompatData<'_>, CompatError> {
    let mut portnum: u32 = 0;
    let mut payload: &[u8] = &[];
    let mut seen_payload = false;

    let mut fields = ProtoFields::new(buf);
    while let Some(item) = fields.next() {
        let (field_no, wire, value) = item?;
        match (field_no, wire) {
            (1, Wire::Varint) => {
                let v = varint(value);
                portnum = u32::try_from(v).unwrap_or(u32::MAX);
            }
            (2, Wire::Len) => {
                if seen_payload {
                    return Err(CompatError::Duplicate);
                }
                seen_payload = true;
                payload = len_bytes(value);
            }
            _ => {}
        }
    }
    if payload.len() > MAX_COMPAT_FRAME_LEN {
        return Err(CompatError::TooLong);
    }
    Ok(CompatData { portnum, payload })
}

/// Protobuf wire types relevant to Meshtastic packets.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Wire {
    Varint,
    Fixed64,
    Len,
    Fixed32,
}

/// A decoded protobuf field value. `Fixed64` values are skipped (RX
/// metadata, unused for display) and never materialized.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum FieldVal<'a> {
    Varint(u64),
    Len(&'a [u8]),
    Fixed32(u32),
}

fn varint(v: FieldVal<'_>) -> u64 {
    match v {
        FieldVal::Varint(n) => n,
        _ => 0,
    }
}

fn fixed32(v: FieldVal<'_>) -> u32 {
    match v {
        FieldVal::Fixed32(b) => b,
        _ => 0,
    }
}

fn len_bytes<'a>(v: FieldVal<'a>) -> &'a [u8] {
    match v {
        FieldVal::Len(b) => b,
        _ => &[],
    }
}

/// Interpret a protobuf int32 varint as signed (internal helper).
fn varint_to_int32(value: u64) -> Result<i32, CompatError> {
    let signed = (value as u64) & 0xFFFF_FFFF_FFFF_FFFF;
    #[allow(clippy::cast_possible_wrap)]
    let as_i64 = signed as i64;
    if as_i64 < i64::from(i32::MIN) || as_i64 > i64::from(i32::MAX) {
        return Err(CompatError::Int32OutOfRange);
    }
    #[allow(clippy::cast_possible_truncation)]
    Ok(as_i64 as i32)
}

/// Zero-allocation protobuf field iterator over a borrowed buffer.
struct ProtoFields<'a> {
    buf: &'a [u8],
    pos: usize,
}

impl<'a> ProtoFields<'a> {
    fn new(buf: &'a [u8]) -> Self {
        Self { buf, pos: 0 }
    }
}

impl<'a> Iterator for ProtoFields<'a> {
    type Item = Result<(u32, Wire, FieldVal<'a>), CompatError>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.pos >= self.buf.len() {
            return None;
        }
        let tag = match read_varint(self.buf, &mut self.pos) {
            Ok(t) => t,
            Err(e) => return Some(Err(e)),
        };
        let field_no = tag >> 3;
        let wire_bits = (tag & 0x07) as u8;
        if field_no == 0 || field_no > u64::from(u32::MAX) {
            return Some(Err(CompatError::BadTag));
        }
        #[allow(clippy::cast_possible_truncation)]
        let field_no = field_no as u32;
        match wire_bits {
            0 => {
                let v = match read_varint(self.buf, &mut self.pos) {
                    Ok(v) => v,
                    Err(e) => return Some(Err(e)),
                };
                Some(Ok((field_no, Wire::Varint, FieldVal::Varint(v))))
            }
            1 => {
                // 64-bit values (RX metadata) are skipped, never stored.
                if self.pos + 8 > self.buf.len() {
                    return Some(Err(CompatError::Truncated));
                }
                self.pos += 8;
                Some(Ok((field_no, Wire::Fixed64, FieldVal::Varint(0))))
            }
            2 => {
                let len = match read_varint(self.buf, &mut self.pos) {
                    Ok(v) => v,
                    Err(e) => return Some(Err(e)),
                };
                if len > self.buf.len() as u64 {
                    return Some(Err(CompatError::Truncated));
                }
                #[allow(clippy::cast_possible_truncation)]
                let len = len as usize;
                if self.pos + len > self.buf.len() {
                    return Some(Err(CompatError::Truncated));
                }
                let bytes = &self.buf[self.pos..self.pos + len];
                self.pos += len;
                Some(Ok((field_no, Wire::Len, FieldVal::Len(bytes))))
            }
            5 => {
                if self.pos + 4 > self.buf.len() {
                    return Some(Err(CompatError::Truncated));
                }
                let mut b = [0u8; 4];
                b.copy_from_slice(&self.buf[self.pos..self.pos + 4]);
                self.pos += 4;
                Some(Ok((
                    field_no,
                    Wire::Fixed32,
                    FieldVal::Fixed32(u32::from_le_bytes(b)),
                )))
            }
            _ => Some(Err(CompatError::UnsupportedWire)),
        }
    }
}

/// Read one protobuf varint at `*pos`, advancing past it (internal helper).
fn read_varint(buf: &[u8], pos: &mut usize) -> Result<u64, CompatError> {
    let mut result: u64 = 0;
    let mut shift: u32 = 0;
    loop {
        if *pos >= buf.len() {
            return Err(CompatError::Truncated);
        }
        let byte = buf[*pos];
        *pos += 1;
        if shift >= 64 {
            return Err(CompatError::VarintOverflow);
        }
        if shift == 63 && byte & 0x7E != 0 {
            // Tenth byte carries only bit 63; higher bits mean >64-bit value.
            return Err(CompatError::VarintOverflow);
        }
        result |= u64::from(byte & 0x7F) << shift;
        if byte & 0x80 == 0 {
            return Ok(result);
        }
        shift += 7;
        if shift > 70 {
            return Err(CompatError::VarintOverflow);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // Hand-built MeshPacket: from=0x12345678 (f1 fixed32), to=broadcast
    // (f2 fixed32), decoded (f4) = Data{portnum=1, payload="Hi"},
    // id=0x2A (f6 fixed32).
    // 0D 78563412 | 15 FFFFFFFF | 22 06 080112024869 | 35 2A000000
    const TEXT_PKT: &[u8] = &[
        0x0D, 0x78, 0x56, 0x34, 0x12, 0x15, 0xFF, 0xFF, 0xFF, 0xFF, 0x22, 0x06, 0x08, 0x01, 0x12,
        0x02, 0x48, 0x69, 0x35, 0x2A, 0x00, 0x00, 0x00,
    ];

    #[test]
    fn decodes_text_packet() {
        let pkt = parse_packet(TEXT_PKT).unwrap();
        assert_eq!(pkt.from_id, 0x1234_5678);
        assert_eq!(pkt.to_id, BROADCAST_ADDR);
        assert_eq!(pkt.packet_id, 0x2A);
        assert!(!pkt.encrypted);
        let data = pkt.data.unwrap();
        assert_eq!(data.portnum, PORT_TEXT);
        assert_eq!(data.text().unwrap(), "Hi");
        assert_eq!(portnum_name(data.portnum), "TEXT_MESSAGE_APP");
    }

    #[test]
    fn encrypted_packet_is_opaque() {
        // f5 (encrypted), 3 bytes: AA BB CC.
        let raw: &[u8] = &[0x2A, 0x03, 0xAA, 0xBB, 0xCC];
        let pkt = parse_packet(raw).unwrap();
        assert!(pkt.encrypted);
        assert_eq!(pkt.encrypted_len, 3);
        assert!(pkt.data.is_none());
    }

    #[test]
    fn rejects_empty_and_noise() {
        assert_eq!(parse_packet(&[]), Err(CompatError::Empty));
        assert!(try_parse_packet(&[0xFF, 0xFF, 0xFF]).is_none());
        // Both decoded and encrypted is rejected.
        let both: &[u8] = &[0x22, 0x00, 0x2A, 0x01, 0x00];
        assert_eq!(
            parse_packet(both),
            Err(CompatError::BothDecodedAndEncrypted)
        );
    }
}
